### ISSUE TITLE
infinity support for Decimal

### ISSUE THREAD

**Issue by zzzeek** [oracle/python-cx_Oracle#78]
cx_Oracle can persist and retrieve the value `float("inf")` but cannot handle `decimal.Decimal("Infinity")`.    I'm marking this as unsupported for now downstream but I do have a user trying to load an infinity value (they can hopefully use floats).

```
import cx_Oracle
import decimal

conn = cx_Oracle.connect(
    user="scott",
    password="tiger",
    dsn=cx_Oracle.makedsn(
        "172.17.0.2", 1521, sid="xe",
    )
)

cursor = conn.cursor()

cursor.execute("""CREATE TABLE t (x BINARY_DOUBLE)""")

try:

    # fails:
    # cursor.execute("insert into t (x) values (:x)",
    #                x=decimal.Decimal("Infinity"))

    # works
    cursor.execute("insert into t (x) values (:x)", x=float("inf"))

    cursor.execute("SELECT x FROM t")
    value = cursor.fetchone()[0]

    # works
    assert value == float("inf")
    cursor.close()

    cursor = conn.cursor()

    def output_type_handler(cursor, name, defaultType,
                            size, precision, scale):
        return cursor.var(
            decimal.Decimal,
            255,
            arraysize=cursor.arraysize)
    cursor.outputtypehandler = output_type_handler
    cursor.execute("SELECT x FROM t")

    # fails:
    # decimal.InvalidOperation: Invalid literal for Decimal: ':000000...'

    print cursor.fetchone()

finally:
    cursor.execute("DROP TABLE t")

```

**Comment by anthony-tuininga** [comment]
Thanks. I'll look into this. I've labelled it an enhancement but it could just as easily be labelled a bug. :-)

**Comment by zzzeek** [comment]
I know I hate those

**Comment by anthony-tuininga** [comment]
Ok. I finally had a chance to look at this. The problem is that the conversion process for decimal.Decimal is to use Oracle type DPI_ORACLE_TYPE_NUMBER and native type DPI_NATIVE_TYPE_BYTES. This won't work for values like infinity since they can't be represented in Oracle's NUMBER format. If instead you use this code

```python
def output_type_handler(cursor, name, defaultType, size, precision, scale):
    return cursor.var(str, arraysize=cursor.arraysize, outconverter=decimal.Decimal)
```

then the conversion will work as expected. I could in theory change decimal.Decimal conversion to use Oracle type DPI_ORACLE_TYPE_VARCHAR, but tests have demonstrated that performance is better if you use DPI_ORACLE_TYPE_NUMBER, so I don't want to change that. You could use the above code in all cases, but the performance isn't as good when using an outconverter either. So I would suggest checking for the defaultType cx_Oracle.NATIVE_FLOAT if you have cases where you want to use decimal.Decimal with native float and possible infinity values.

**Comment by zzzeek** [comment]
> but the performance isn't as good when using an outconverter either

yeah that's part of why this whole thing in SQLAlchemy, especially the rewrite I've done for 1.2, is crazy, to accommodate for a very complicated set of conditionals

> So I would suggest checking for the defaultType cx_Oracle.NATIVE_FLOAT if you have cases where you want to use decimal.Decimal with native float and possible infinity values.

OK so...here's where the whole function is at, at the moment - "self.asdecimal" means they want numerics back as Decimal and false means they want float():

        def handler(cursor, name, default_type, size, precision, scale):
            outconverter = None
            if precision:
                if self.asdecimal:
                    if is_cx_oracle_6:
                        type_ = decimal.Decimal
                    else:
                        type_ = cx_Oracle.STRING
                        outconverter = dialect._to_decimal
                else:
                    if self.is_number and scale == 0:
                        if has_native_int:
                            type_ = cx_Oracle.NATIVE_INT
                        else:
                            type_ = cx_Oracle.NUMBER
                            outconverter = int
                    else:
                        type_ = cx_Oracle.NATIVE_FLOAT
            else:
                if self.asdecimal:
                    if is_cx_oracle_6:
                        type_ = decimal.Decimal
                    else:
                        type_ = cx_Oracle.STRING
                        outconverter = dialect._to_decimal
                else:
                    if self.is_number and scale == 0:
                        if has_native_int:
                            type_ = cx_Oracle.NATIVE_INT
                        else:
                            type_ = cx_Oracle.NUMBER
                            outconverter = int
                    else:
                        type_ = cx_Oracle.NATIVE_FLOAT

            return cursor.var(
                type_, 255,
                arraysize=cursor.arraysize,
                outconverter=outconverter
            )

        return handler

so what you are saying is, both of the conditionals where they want a decimal and I'm cheking is_cx_oracle_6 and so happy to not use "outconverter" anymore, those have to go...right?   because what else am I going to get as the value of default_type besides NATIVE_FLOAT for numeric types?

**Comment by zzzeek** [comment]
also my _to_decimal thing does that madness where they've changed their decimal point character...so yeah I'm probably not going to turn this on by default.  there needs to be some hook or something for this unbelievably unlikely use case

**Comment by anthony-tuininga** [comment]
> so what you are saying is, both of the conditionals where they want a decimal and I'm cheking is_cx_oracle_6 and so happy to not use "outconverter" anymore, those have to go...right? because what else am I going to get as the value of default_type besides NATIVE_FLOAT for numeric types?

I'm not sure I understand your question completely. I'll try to answer and you can correct my misunderstandings!

The value of default_type can be cx_Oracle.NUMBER as well, and then the conversion to decimal will work just fine because infinity isn't possible with Oracle Number. But if the default_type is NATIVE_FLOAT you want to use the out converter in that case only. Hope that makes sense!

> also my _to_decimal thing does that madness where they've changed their decimal point character...so yeah I'm probably not going to turn this on by default. there needs to be some hook or something for this unbelievably unlikely use case

Are you referring to your code? Or something else? I'm not sure what you were trying to say. Can you clarify?

**Comment by zzzeek** [comment]
> The value of default_type can be cx_Oracle.NUMBER as well, and then the conversion to decimal will work just fine because infinity isn't possible with Oracle Number. But if the default_type is NATIVE_FLOAT you want to use the out converter in that case only.

yes that was all it was, that I may see NATIVE_FLOAT or I may see NUMBER.     The issue at the moment is I spent weeks rewriting all this code to both modernize for cx_Oracle 5 and 6 and I was hoping to not have to keep tweaking it, so I'm just going to make these changes in some future round.

> Are you referring to your code? Or something else? I'm not sure what you were trying to say. Can you clarify?

this refers to my code but also the reality of when you use an outconverter with numerics.  It is not sufficient to say:

   outconverter=decimal.Decimal

because if the NLS_TERRITORY environment variable is set to something like "GERMANY", the string that is passed to decimal.Decimal contains a comma inside of it, not a decimal point, and the constructor will fail.   SQLAlchemy's dialect, when it uses an outconverter for decimals, instead uses a function _to_decimal which is set up upon first connect based on examining the NLS_NUMERIC_CHARACTERS parameter, and if the character is not a ".", then a more expensive function is swapped in which accommodates for the comma (or whatever character) before sending off to Decimal.

Basically this is stuff that really needs to be handled by the DBAPI.   cx_Oracle 6 gets much closer to that except for this infinity issue.

**Comment by anthony-tuininga** [comment]
Hmm, I have a better solution, I believe! If the default type is cx_Oracle.NATIVE_FLOAT, just continue to use that type with the outconverter function set to decimal.Decimal. That will bring the floating point number directly over to cx_Oracle and will pass that value directly to the decimal.Decimal constructor -- which will happily convert the floating point number to decimal.Decimal. That works in 5.x and 6.x equally well (I just confirmed with 5.2, 5.3 and 6). And because it doesn't look at NLS_NUMERIC_CHARACTERS there is no concerns about that value either.

If the default type is cx_Oracle.NUMBER, then in cx_Oracle 6 you can use the type decimal.Decimal directly without any further concerns. That also doesn't concern itself with NLS_NUMERIC_CHARACTERS.

To summarise: if you add one if clause at the beginning of your function that does this:

```python
if defaultType == cx_Oracle.NATIVE_FLOAT and self.asdecimal:
    type_ = defaultType
   outConverter = decimal.Decimal
elif precision:
    do_the_rest_of_the function
```
that should solve your problems with infinity. What do you think?

**Comment by anthony-tuininga** [comment]
I believe this has been resolved in the best way it can be. Feel free to re-open if you disagree, however!

**Comment by zzzeek** [comment]
yes I will try out the final suggestion you have and also add back the tests for infinity to see if they pass.  thanks!

**Comment by zzzeek** [comment]
oh and also we can't have INSERT here still.    

```
cursor.execute(
    'INSERT INTO t1 (intcol, numericcol) VALUES (:intcol, :numericcol)', 
     {'intcol': 1, 'numericcol': Decimal('Infinity')})

"TypeError: an integer is required"
```

**Comment by anthony-tuininga** [comment]
What is the structure of table t1?

**Comment by zzzeek** [comment]
that's 5.0.x.  in 6.0.x I get "DPI-1043: invalid number".    getting you the table...

**Comment by zzzeek** [comment]
```
CREATE TABLE t1 (
	intcol INTEGER, 
	numericcol BINARY_DOUBLE
)


INSERT INTO t1 (intcol, numericcol) VALUES (:intcol, :numericcol)
{'intcol': 1, 'numericcol': Decimal('Infinity')}

cx_Oracle.DatabaseError: DPI-1043: invalid number

```

**Comment by zzzeek** [comment]
im not sure there's a way around that, includingon your end, other than by "if number == Decimal("Infinity")" and that is going to be slow.    I dont htink this case is worth worrying over

**Comment by zzzeek** [comment]
I have it doing "if value.is_infinite()" as the test, this is fine.   we're done.

**Comment by anthony-tuininga** [comment]
Great. Thanks!
